### ISSUE TITLE
Better MessageBox

### ISSUE THREAD

**Issue by Andrej123456789** [dotnet/wpf#5967]
MessageBox is design is kind of Windows 9.x era.
Please add better design at MessageBox.

Example - WPF:
![image](https://user-images.githubusercontent.com/83548580/150091734-56dcf68b-9d09-46fa-8299-aaffa4c263fd.png)

Example - WinForms:
![image](https://user-images.githubusercontent.com/83548580/150092217-c7f7e23d-c40b-4e32-a038-3e7919cb4c15.png)

WinForms's MessageBox is much better.

**Comment by lindexi** [comment]
But we can use the WinForms's MessageBox in WPF.

**Comment by ThomasGoulet73** [comment]
@Andrej123456789 You need an app mainfest to use the more modern controls.

Content of app.manifest:
```
<?xml version="1.0" encoding="utf-8"?>
<assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
  <dependency>
    <dependentAssembly>
      <assemblyIdentity
          type="win32"
          name="Microsoft.Windows.Common-Controls"
          version="6.0.0.0"
          processorArchitecture="*"
          publicKeyToken="6595b64144ccf1df"
          language="*"
        />
    </dependentAssembly>
  </dependency>
</assembly>
```

WinForms seems to work without the need for an app manifest (My guess is that it comes from this line: https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L388). It would be interesting to try to add in WPF the code from WinForms that fixes this.

**Comment by ClosetBugSlayer** [comment]
WPF should be using TaskDialog not MessageBox. Every supported version of Windows has this API now.

**Comment by psmulovics** [comment]
https://docs.microsoft.com/en-us/windows/win32/controls/task-dialogs

**Comment by RussKie** [comment]
Windows Forms SDK provides an [app.manifest by default](https://github.com/dotnet/winforms/blob/12d87648900ba1668c0bad7e63c4cba8f58c5abb/src/System.Windows.Forms/src/System/Windows/Forms/Application.cs#L835-L837), so the comctl 6.0 is used by default as well (with the ubiquitous `Appplication.UseVisualStyles()` call in Program.cs).
I don't know if WPF provides a default manifest, so this is something to check.

**Comment by RussKie** [comment]
> WPF should be using TaskDialog not MessageBox. Every supported version of Windows has this API now.

It isn't straightforward to correctly use Win32 API, and it requires a lot of elbow grease without managed wrapper (e.g., https://github.com/dotnet/winforms/pull/1133).

**Comment by Andrej123456789** [comment]
I will try it soon

**Comment by ClosetBugSlayer** [comment]
> It isn't straightforward to correctly use Win32 API, and it requires a lot of elbow grease without managed wrapper (e.g., [dotnet/winforms#1133](https://github.com/dotnet/winforms/pull/1133)).

The only elbow grease is if you use an activation context, which I do personally just to be safe. Otherwise the default manifest should pull in the API as-is; it works for me in .NET 6 at least.

**Comment by Andrej123456789** [comment]
> @Andrej123456789 MessageBox is just a wrapper around the Win32 control. WPF does not control the styling inside the MessageBox. You need an app mainfest to use the more modern controls.
> 
> Content of app.manifest:
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> <assembly manifestVersion="1.0" xmlns="urn:schemas-microsoft-com:asm.v1">
>   <dependency>
>     <dependentAssembly>
>       <assemblyIdentity
>           type="win32"
>           name="Microsoft.Windows.Common-Controls"
>           version="6.0.0.0"
>           processorArchitecture="*"
>           publicKeyToken="6595b64144ccf1df"
>           language="*"
>         />
>     </dependentAssembly>
>   </dependency>
> </assembly>
> ```
> 
> WinForms seems to work without the need for an app manifest (My guess is that it comes from this line: https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L388). It would be interesting to try to add in WPF the code from WinForms that fixes this.

I tried this but I can't compile project anymore.  [Here](https://gist.github.com/Andrej123456789/3b692446ef4a05e36a90dd6484309a82) is file so you can check it out.

![image](https://user-images.githubusercontent.com/83548580/154852303-2b4689f5-64d1-4fe8-9d35-2ada6b757697.png)
Here is screenshot of error

**Comment by Andrej123456789** [comment]
Nevermind, now works

**Comment by Andrej123456789** [comment]
But point of this issue to use TaskDialog as default

**Comment by miloush** [comment]
@Andrej123456789 no the original issue is asking for WinForms matching message box, and that is solved using manifest. If you have a TaskDialog API request, upvote #2902.

**Comment by Andrej123456789** [comment]
Oh

**Comment by aquinn39** [comment]
Another way to solve this if you can't/don't want to use the app manifest is to call System.Windows.Forms.Application.EnableVisualStyles() when the app starts. Similar to the manifest solution, this enables visual styles for Forms and GDI controls, such as the message box.

**Comment by lindexi** [comment]
@aquinn39 By default, we will reference the System.Windows.Forms in dotnet core.

**Comment by dipeshmsft** [comment]
@Andrej123456789, @miloush got a bit confused about what are we talking about in this issue.

TaskDialog request is a different issue, I am clear on that.

Now, going through the first comment in this thread, it's mentioned that "Please add better design at MessageBox."

1. Does this refer to the rounded corners ( as in Windows 11 theme ) ??
2. Or are we referring to the icon and buttons in the message box control ??

Because we have other ```public Show``` functions which can be used to display those. https://github.com/dotnet/wpf/blob/a71d94d93b4af94061f6cd27c57b1d00efaccaab/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/MessageBox.cs#L95-L146

**Comment by miloush** [comment]
@dipeshmsft I agree that the original post does not actually mention what the perceived issue is. From the screenshot, my interpretation was the button styling, which can be solved by manifest. Good call pointing out that the user can add an icon and a title with different overload.

**Comment by Andrej123456789** [comment]
> @Andrej123456789, @miloush got a bit confused about what are we talking about in this issue.
> 
> TaskDialog request is a different issue, I am clear on that.
> 
> Now, going through the first comment in this thread, it's mentioned that "Please add better design at MessageBox."
> 
> 1. Does this refer to the rounded corners ( as in Windows 11 theme ) ??
> 2. Or are we referring to the icon and buttons in the message box control ??
> 
> Because we have other `public Show` functions which can be used to display those.
> 
> https://github.com/dotnet/wpf/blob/a71d94d93b4af94061f6cd27c57b1d00efaccaab/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/MessageBox.cs#L95-L146

There are already rounded corners so I my focus is on buttons

**Comment by victorgoebel** [comment]
You can use the **[Task Dialog](https://github.com/ookii-dialogs/ookii-dialogs-wpf)** for more modern Message Boxes:

![image](https://user-images.githubusercontent.com/73068340/202356338-a7fb06a9-7ec1-4b93-b02f-0436c75a1648.png)

![image](https://user-images.githubusercontent.com/73068340/202356347-bb105f6b-5934-4077-ad49-71b5054820a9.png)

**Comment by RussKie** [comment]
FYI TaskDialog requires comctl v6

**Comment by miloush** [comment]
@RussKie which ships with Vista so it should be fine. With CreateActCtx you don't even need a manifest, though I am wondering why that is not part of the default embedded manifest nowadays.

**Comment by Andrej123456789** [comment]
Do you have planes for changing default manifest to use modern controls

**Comment by snechaev** [comment]
> WinForms seems to work without the need for an app manifest (My guess is that it comes from this line: https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L388). It would be interesting to try to add in WPF the code from WinForms that fixes this.

The link is outdated. Here is the correct permanent link to the specific revision.: https://github.com/dotnet/winforms/blob/b7c5ac63cb8432f7bdae8474adffc2b3a37d0ea1/src/System.Windows.Forms/src/System/Windows/Forms/MessageBox.cs#L388
